workflowRDF/fairworkflow.py

```diff
@@ -76,69 +139,26 @@ def __str__(self):
     def rdf_to_file(self, fname, format='turtle'):
         return self.get_rdf().serialize(destination=fname, format=format)
         
-    def nanopublish(self, url=None):
+    def nanopublish(self):
+        # Publish all the steps individually
         for step in self.steps:
-            step.nanopublish(url=url)
+            step.nanopublish()
+
+        # Publish the workflow itself
+        nanopublish(assertionrdf=self.get_rdf(), uri=self.np_uri)
+
 
 class FairStepEntry:
-    def __init__(self, rdf_node, func, args, kwargs, rdf_graph):
-        self.this_step = rdf_node
+    def __init__(self, func, args, kwargs):
         self.func = func
         self.args = args
         self.kwargs = kwargs
-        self.rdf = rdf_graph
         self.executed = False
         self.result = None
 
-
-    def nanopublish(self, url=None):
-
-        np_uri = "http://www.example.org/pub1"
-
-        THIS = rdflib.Namespace(np_uri)
-        SUB = rdflib.Namespace(np_uri+"#")
-
-        # Set up different contexts
-        np_rdf = rdflib.ConjunctiveGraph()
-        head = rdflib.Graph(np_rdf.store, SUB.head)
-        assertion = rdflib.Graph(np_rdf.store, SUB.assertion)
-        provenance = rdflib.Graph(np_rdf.store, SUB.provenance)
-        pubInfo = rdflib.Graph(np_rdf.store, SUB.pubInfo)
-
-        np_rdf.bind("this", THIS)
-        np_rdf.bind("sub", SUB)
-        np_rdf.bind("np", NP)
-
-        np_rdf.bind("p-plan", PPLAN)
-        np_rdf.bind("plex", PLEX)
-        np_rdf.bind("edam", EDAM)
-        np_rdf.bind("prov", PROV)
-        np_rdf.bind("dul", DUL)
-        np_rdf.bind("bpmn", BPMN)
-        np_rdf.bind("pwo", PWO)
-        np_rdf.bind("rdfg", RDFG)
-
-
-
-        head.add((THIS[''], RDF.type, NP.Nanopublication))
-        head.add((THIS[''], NP.hasAssertion, SUB.assertion))
-        head.add((THIS[''], NP.hasProvenance, SUB.provenance))
-        head.add((THIS[''], NP.hasPublicationInfo, SUB.pubInfo))
-
-        assertion += self.rdf # A (temporary) misuse of nanopublications
-
-        creationtime = rdflib.Literal(datetime.now(),datatype=XSD.date)
-        provenance.add((SUB.assertion, PROV.generatedAtTime, creationtime))
-        provenance.add((SUB.assertion, PROV.wasDerivedFrom, THIS.experiment)) 
-        provenance.add((SUB.assertion, PROV.wasAttributedTo, THIS.experimentScientist))
-
-        pubInfo.add((THIS[''], PROV.wasAttributedTo, THIS.DrBob))
-        pubInfo.add((THIS[''], PROV.generatedAtTime, creationtime))
-
-        # Convert nanopub rdf to trig
-        stepname = str(self.func).replace(' ', '')
-        fname = f'step_{stepname}.trig'
-        np_rdf.serialize(destination=fname, format='trig')
+        self.np_uri = "http://purl.org/nanopub/temp/FAIRWorkflowsTest/Step"
+        self.THISSTEP = rdflib.Namespace(self.np_uri + '#')
```

Comment:
What's the difference between `self.THISSTEP` and `self.STEP`? Also, I would not put these variable names in capitals if they're not constants.

Reply:
This is to do with the horror of linking nanopubs together with the correct references, while allowing the code to produce RDF for non-nanopub reasons. Basically, when you construct the workflow RDF, you need a URI by which to refer to the individual steps. Normally you can just use the THISSTEP namespace, and be done with it. However, if you want to publish these as nanopubs, then the hashing/signing process gives the nanopub a new, special reference. This means the workflow RDF needs to know to use the nanopublished URI if/when an individual step has been published. This is what is being handled by STEP above (it is initially equal to THISSTEP, and when published, becomes equal to the URI of its corresponding nanopublication). I agree that this is confusing and will tidy it up.